docs/how-to/type-mapping.rst

```diff
@@ -119,45 +177,157 @@ string::
+    7
+
+.. note::
+
+    Python objects stored in an :class:`~rubicon.objc.api.NSArray` are
+    converted to Objective-C objects using the rules described in
+    :ref:`argument_conversion`.
 
 Dictionaries
-------------
-
-If a method calls for an :class:`NSDictionary` or :class:`NSMutableDictionary`
-argument, you can provide a Python :class:`dict`. Rubicon will construct an
-:class:`NSMutableDictionary` instance from the data in the :class:`dict`
-provided, and pass that value for the argument.
-
-If a method returns an :class:`NSDictionary` or :class:`NSMutableDictionary`,
-the return value will be a wrapped :class:`ObjCDictInstance` type. This type
-implements a :class:`dict`-like interface, wrapped around the underlying
-:class:`NSDictionary` data. This means you can treat the return value as if it
-were a dict - iterating over keys, values or items, retrieving objects by key,
-and so on.
-
-
-:class:`NSPoint`, :class:`NSSize`, and :class:`NSRect`
-------------------------------------------------------
-
-On instances of an Objective C structure, each field is exposed as a Python
-attribute. For example, if you create an instance of an :class:`NSSize` object
-you can access its width and height by calling :meth:`NSSize.width`.
+""""""""""""
+
+:class:`~rubicon.objc.api.NSDictionary` objects behave like any other Python
+mapping - their items can be accessed and standard operations like :func:`len`
+and ``in`` are supported:
+
+.. code-block:: python
+
+    >>> from rubicon.objc import NSDictionary
+    >>> d = objc.NSDictionary.dictionaryWithDictionary({"one": 1, "two": 2})
+    >>> d["one"]
+    1
+    >>> len(d)
+    >>> 2
+    >>> "two" in d
+    True
+    >>> "five" in d
+    False
 
-When you need to pass an Objective C structure to an Objective C method,
-you can pass a tuple instead. For example, if you pass (10.0, 5.1) where a
-:class:`NSSize` is expected, it will be converted automatically in the appropriate
-width, height for the structure.
+.. note::
+
+    :class:`~rubicon.objc.api.ObjCInstance` objects wrapping a
+    :class:`~rubicon.objc.api.NSDictionary` internally have the class
+    ``ObjCDictInstance`` or ``ObjCMutableDictInstance``, and you will see these
+    names in the :func:`repr` of :class:`~rubicon.objc.api.NSDictionary`
+    objects. This is an implementation detail - you should not refer to the
+    ``ObjCDictInstance`` and ``ObjCMutableDictInstance`` classes explicitly in
+    your code.
+
+:class:`~rubicon.objc.api.NSMutableDictionary` objects additionally support
+mutating operations, like item assignment:
+
+.. code-block:: python
+
+    >>> md = objc.NSMutableDictionary.dictionaryWithDictionary({"one": 1, "two": 2})
+    >>> md["three"] = 3
+    >>> md
+    <rubicon.objc.collections.ObjCMutableDictInstance 0x10b8a7860: __NSDictionaryM at 0x7f86fb164b60: {
+        one = 1;
+        three = 3;
+        two = 2;
+    }>
+
+Mapping methods like ``keys`` and ``values`` are also supported:
```

Comment:
> Do we want to mention that these create a copy of the key/value list, rather than a MemoryView?

I've added a note about this to the reference docs, because I think in most cases it's unlikely to be relevant in practice - you don't often call `keys`/`values` repeatedly, and most dictionaries will be small enough that there's no noticeable performance impact.

> To that end - the docs for str, list and dict handling talk about converting from Python to ObjC, but don't mention how to convert from ObjC to a pure Python str/list/dict should you need to

You're right, that's not explained very well in the howto or in the reference documentation - it's only explicitly pointed out for strings, because of the UTF-16 thing. I'm not sure how often it's really necessary to convert an Objective-C collection to a "real" Python one (since our wrapper classes already implement the Python collection APIs), but I agree that it should be mentioned for completeness. I'll expand both the how-to and reference docs to include that.